packages/swing-store/src/swingStore.js

```diff
@@ -311,69 +373,179 @@ function makeSwingStore(dirPath, forceReset, options) {
-    close: closeStreamStore,
-    ...streamStore
-  } = sqlStreamStore(dirPath, { sqlite3 });
+  const kernelKVStore = {
+    ...kvStore,
+    set(key, value, bypassHash) {
+      assert.typeof(key, 'string');
+      const keyType = getKeyType(key);
+      assert(keyType !== 'host');
+      set(key, value);
+      if (keyType === 'consensus' && !bypassHash) {
+        crankhasher.add('add');
+        crankhasher.add('\n');
+        crankhasher.add(key);
+        crankhasher.add('\n');
+        crankhasher.add(value);
+        crankhasher.add('\n');
+      }
+    },
+    delete(key) {
+      assert.typeof(key, 'string');
+      const keyType = getKeyType(key);
+      assert(keyType !== 'host');
+      del(key);
+      if (keyType === 'consensus') {
+        crankhasher.add('delete');
+        crankhasher.add('\n');
+        crankhasher.add(key);
+        crankhasher.add('\n');
+      }
+    },
+  };
+
+  const hostKVStore = {
+    ...kvStore,
+    set(key, value) {
+      const keyType = getKeyType(key);
+      assert(keyType === 'host');
+      set(key, value);
+    },
+    delete(key) {
+      const keyType = getKeyType(key);
+      assert(keyType === 'host');
+      del(key);
+    },
+  };
 
-  const snapshotDir = path.resolve(dirPath, 'xs-snapshots');
-  fs.mkdirSync(snapshotDir, { recursive: true });
-  const snapStore = makeSnapStore(snapshotDir, makeSnapStoreIO(), {
-    keepSnapshots,
-  });
+  const streamStore = makeSQLStreamStore(db, ensureTxn);
+  let snapStore;
+
+  if (dirPath) {
+    const snapshotDir = path.resolve(dirPath, 'xs-snapshots');
+    fs.mkdirSync(snapshotDir, { recursive: true });
+    snapStore = makeSnapStore(snapshotDir, makeSnapStoreIO(), {
+      keepSnapshots,
+    });
+  }
+
+  const savepoints = [];
+  const sqlReleaseSavepoints = db.prepare('RELEASE SAVEPOINT t0');
+
+  function startCrank() {
+    assert(!inCrank);
+    inCrank = true;
+    resetCrankhash();
+  }
 
-  /** @param {boolean} [abort] */
-  function doCommit(abort) {
-    if (!txnFinish) {
-      return undefined;
+  function establishCrankSavepoint(savepoint) {
+    assert(inCrank);
+    const savepointOrdinal = savepoints.length;
+    savepoints.push(savepoint);
+    const sql = db.prepare(`SAVEPOINT t${savepointOrdinal}`);
+    sql.run();
+  }
+
+  function rollbackCrank(savepoint) {
+    assert(inCrank);
+    for (const savepointOrdinal of savepoints.keys()) {
+      if (savepoints[savepointOrdinal] === savepoint) {
+        const sql = db.prepare(`ROLLBACK TO SAVEPOINT t${savepointOrdinal}`);
+        sql.run();
+        savepoints.length = savepointOrdinal;
+        return;
+      }
     }
-    txnFinish(abort ? lmdbAbort : undefined);
-    return Promise.resolve(txnDone)
-      .then(() => {
-        trace(`${abort ? 'abort' : 'commit'}-tx`);
-      })
-      .finally(() => {
-        txnDone = null;
-        txnFinish = null;
-      });
+    assert.fail(`no such savepoint as "${savepoint}"`);
+  }
+
+  function emitCrankHashes() {
```

Comment:
It *is* doing IO.  It's not just computing the hashes, it's writing them to the DB.  I like "emit" because it captures both the computation and the output.